Feature/nls (#914)
MicroEJ/docs: ApplicationDeveloperGuide/nls.rst

   String label = Labels.NLS.getMessage(Labels.Label1);

> Default Locale

Reviewer: Maybe more simply `Locale` because it's about the current locale and the default locale.
```suggestion
Locale
```

--------------

For the application to know which language to use among those made available and when, you can set it and change it at any point using the `setCurrentLocale(locale)`_ method. 
If no locale has been set yet when getting a message, the translation for the first locale available in alphabetical order will be used by default. 
However, you can also pick this locale to default to yourself, by adding a ``com.microej.binarynls.defaultLocale`` property followed by a locale name in a ``.properties.list`` file. 

.. _binary-nls: https://repository.microej.com/modules/com/microej/library/runtime/binary-nls
.. _setCurrentLocale(locale): https://repository.microej.com/javadoc/microej_5.x/apis/ej/nls/NLS.html#setCurrentLocale-java.lang.String-

Plural Forms
------------

> Starting with version 4.0.0 of the `NLS module`_ and version 3.0.0 of the `binary-nls`_ module is introduced support of GNU gettext's plural form feature in PO files.

Reviewer: ```suggestion
Starting with version 4.0.0 of the `NLS module`_ and version 3.0.0 of the `binary-nls`_ module is introduced the support of GNU gettext's plural form feature in PO files.
```

In the case of a message with plural forms in PO files, this works much the same way, using the messages and forms in the fallback locale if available.
If no fallback locale is specified or if the requested message is not specified in it, then the ``msgid`` will be used for a count value of 1, and the ``msgid_plural`` will be used for any other value, as gettext would function.

.. _ej.nls.NLS.getMessage(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/nls/NLS.html#getMessage-int-

Converter
---------

Problematic
^^^^^^^^^^^

> Translations can be used directly for the following purposes:

Reviewer: ```suggestion
Translated messages can be used directly for the following purposes:
```
or something like this.

* EDC (in the console): `System.out.println()`_.
* MicroUI: `ej.microui.display.Painter.drawString()`_.
* MicroVG: `ej.microvg.VectorGraphicsPainter.drawString()`_.

When displaying certain languages, such as Arabic, string analysis is necessary for character substitution and right-to-left (RTL) reading direction. 
:ref:`Console encoding <set_console_encoding>` is required for proper display using EDC.

.. tabs::

   .. tab:: Without Console Encoding

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         'D91(J)

   .. tab:: With Console Encoding

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         العربية

> In order to display a message with MicroVG, the :ref:`complex layout <section_vg_font_complex>` must be used. 

Reviewer: ```suggestion
In order to display correctly this kind of message with MicroVG, the :ref:`complex layout <section_vg_font_complex>` must be used. 
```
or `such a message` like below.

.. tabs::

   .. tab:: Without Complex Layout

      .. figure:: UI/NLS/images/microvg_not_converted_simple.png

   .. tab:: With Complex Layout

      .. figure:: UI/NLS/images/microvg_not_converted_complex.png

It's not possible to display such a message with MicroUI: the Graphics Engine doesn't offer substitution table reading or bidirectional string management.
The rendering is systematically wrong:

.. figure:: UI/NLS/images/microui_not_converted.png

.. _System.out.println(): https://repository.microej.com/javadoc/microej_5.x/apis/java/io/PrintStream.html#println--
.. _ej.microui.display.Painter.drawString(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/Painter.html#drawString-ej.microui.display.GraphicsContext-java.lang.String-ej.microui.display.Font-int-int-
.. _ej.microvg.VectorGraphicsPainter.drawString(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microvg/VectorGraphicsPainter.html#drawString-ej.microui.display.GraphicsContext-java.lang.String-ej.microvg.VectorFont-float-float-float-

Solution
^^^^^^^^

Since the version 3.1.0, `binary-nls`_ module features an offboard translation conversion.

> The generated string has already been substituted and rearranged.

Reviewer: ```suggestion
It means that the generated strings can be already substituted and rearranged.
```

This conversion enables MicroUI's Graphics Engine to render complex strings correctly.

.. warning:: This offboard conversion only concerns PO files.

.. tabs::

   .. tab:: Without Offboard Conversion

      .. figure:: UI/NLS/images/microui_not_converted.png

   .. tab:: With Offboard Conversion

      .. figure:: UI/NLS/images/microui_converted.png

This also avoids embedding substitution tables and the complex layout management when the message is displayed with MicroVG.

Principle
^^^^^^^^^

Keep in mind that offboard conversion is only relevant to translation values. 

> It is important to note that all other fields, such as message identifiers and display names, are not converted as they are not intended to be displayed.

Reviewer: Mmh, I do not agree that the display names are not used to be displayed (hence the name of the property).
What we say is that it is not yet supported but we will see later if it is necessary. Maybe just remove this line?

Author: Only messages are converted, not other fields, and they will never be converted...

Reviewer: Yep, but still, I think we cannot say that the "display name" is not intended to be "displayed" IMHO.

Example: العربية

* Unicodes: :guilabel:`U+627` :guilabel:`U+644` :guilabel:`U+639` :guilabel:`U+631` :guilabel:`U+628` :guilabel:`U+64a` :guilabel:`U+629`
* Text shaping: :guilabel:`U+fe8d` :guilabel:`U+fedf` :guilabel:`U+fecc` :guilabel:`U+feae` :guilabel:`U+fe91` :guilabel:`U+fef4` :guilabel:`U+fe94`
* Redirection: :guilabel:`U+fe94` :guilabel:`U+fef4` :guilabel:`U+fe91` :guilabel:`U+feae` :guilabel:`U+fecc` :guilabel:`U+fedf` :guilabel:`U+fe8d`

Hebrew
""""""

This converter is dedicated to the Hebrew language, which involves bidirectional reordering of text. 

* Example  xxx_todo

Limitations
^^^^^^^^^^^

Conversion is a feature dedicated to graphic display (MicroUI or MicroVG).
A message converted and displayed with :ref:`EDC <set_console_encoding>` may be displayed incorrectly, especially directionality.

.. tabs::

   .. tab:: Without Offboard Conversion

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         العربية

   .. tab:: With Offboard Conversion

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         ﺔﻴﺑﺮﻌﻟﺍ

> Messages are usually displayed using a single type of display, either EDC or UI. 

Reviewer: ```suggestion
Messages are usually displayed using a single type of medium, either EDC or UI. 
```
or `output` or `target` or …

> However, if a dual display is required, the PO file must be explicitly tagged as *to be converted* (see above) to ensure compatibility with the UI display. 

Reviewer: ```suggestion
However, if several [same as previous comment] is required, the PO file must be explicitly tagged as *to be converted* (see above) to ensure compatibility with the UI display. 
```

Author: There are a lot of locations where the word `display` is used to talk about the rendering (console or UI). I know it is problematic but the word for "affichage" is "display". 

@seonmej @microej-sh I'll let you correct the wording

Reviewer: I am ok with the reword. `dual display` was ambiguous for me.

> When displaying it, the application needs to add the character :guilabel:`U+202D` before the message to force the message orientation.

Reviewer: ```suggestion
But when displaying it with EDC print, the application needs to add the character :guilabel:`U+202D` before the message to force the message orientation.
```

* Redirection: :guilabel:`U+fe94` :guilabel:`U+fef4` :guilabel:`U+fe91` :guilabel:`U+feae` :guilabel:`U+fecc` :guilabel:`U+fedf` :guilabel:`U+fe8d`

Hebrew
""""""

This converter is dedicated to the Hebrew language, which involves bidirectional reordering of text.

Niqqud is currently not supported.

* Example:

.. code-block:: po

   msgid ""
   msgstr ""
   "Language: he\n"
   "Language-Team: Hebrew\n"
   "Language-Converter: Hebrew\n"
   "MIME-Version: 1.0\n"
   "Content-Type: text/plain; charset=UTF-8\n"

   msgid "Hello"
   msgstr "שלום"

Limitations
^^^^^^^^^^^

Conversion is a feature dedicated to graphic display (MicroUI or MicroVG).
A message converted and displayed with :ref:`EDC <set_console_encoding>` may be shown incorrectly, especially regarding visual orientation.

.. tabs::

   .. tab:: Without Offboard Conversion

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         العربية

   .. tab:: With Offboard Conversion

      .. code-block:: java

         System.out.print("العربية");

      .. code-block:: console

         ﺔﻴﺑﺮﻌﻟﺍ

> Messages are usually displayed using a single type of output, either EDC or UI.
> To properly render the text on the UI display, the PO file must explicitly specify a converter (see above) to ensure compatibility.
> But when printing it with EDC, the application needs to add the character :guilabel:`U+202D` before the message to force the message orientation, and :guilabel:`U+202C` after it to restore the previous orientation.

Reviewer: ```suggestion
Messages are usually displayed using a single type of output, either EDC or UI.

When printing the text with EDC, it is correctly rendered without any pre-conversion (the terminal on the PC, that actually prints the text, performs the necessary reordering, substitutions, etc.)

To properly render the text on the UI display, the PO file must explicitly specify a converter (see above) to ensure compatibility.
But when printing a pre-converted text with EDC, the application needs to add the character :guilabel:`U+202D` before the message to force the message orientation, and :guilabel:`U+202C` after it to restore the previous orientation.
```
